dart_application_3/bin/task.dart

```diff
@@ -0,0 +1,211 @@
+      return false;
+    }
+
+    nodeMap[parentNode]!.childrenList.add(childNode);
+    nodeMap[childNode]!.parentList.add(parentNode);
+    print("dependnecy is added");
+    return true;
+  }
+
+  /**
+   * create node
+   * validate node
+   * add node
+   */
+  static void addNode(Map<int, Node> nodeMap, int nodeId, String nodeName) {
+    List<int> childrenList = [];
+    List<int> parentList = [];
+    Node newNode = new Node(
+        nodeId: nodeId,
+        nodeName: nodeName,
+        childrenList: childrenList,
+        parentList: parentList);
+    nodeMap[nodeId] = newNode;
+  }
+
+  /**
+   * check if dependency exist
+   */
+  static bool _dependency(
+      int parentNode, int childNode, Map<int, Node> nodeMap) {
+    bool exist = nodeMap[parentNode]!.childrenList.contains(childNode);
+    return exist;
+  }
+
+  /**
+   * bfs traversal for ancestors
+   */
+  static List<int> _searchAncestors(Map<int, Node> nodeMap, int node) {
+    Queue<int> nodeQueue = new Queue();
+    nodeQueue.add(node);
+    nodeMap[node]!.visited = true;
+    List<int> ancestorList = [];
+    while (nodeQueue.isNotEmpty) {
+      int nodeAncestor = nodeQueue.removeLast();
+      if (nodeAncestor != node) ancestorList.add(nodeAncestor);
+      for (int link in nodeMap[nodeAncestor]!.parentList) {
+        if (!nodeMap[link]!.visited) {
+          nodeMap[link]!.visited = true;
+          nodeQueue.add(link);
+        }
+      }
+    }
+    nodeMap.forEach((node, value) {
+      value.visited = false;
+    });
+    return ancestorList;
+  }
+
+  /**
+   * bfs traversal for descendents
+   */
+  static List<int> _searchDescendents(Map<int, Node> nodeMap, int node) {
+    Queue<int> nodeQueue = new Queue();
+    nodeQueue.add(node);
+    nodeMap[node]!.visited = true;
+    List<int> descendentList = [];
+    while (nodeQueue.isNotEmpty) {
+      int nodeAncestor = nodeQueue.removeLast();
+      if (nodeAncestor != node) descendentList.add(nodeAncestor);
+      for (int link in nodeMap[nodeAncestor]!.childrenList) {
+        if (!nodeMap[link]!.visited) {
+          nodeMap[link]!.visited = true;
+          nodeQueue.add(link);
+        }
+      }
+    }
+    nodeMap.forEach((node, value) {
+      value.visited = false;
+    });
+    return descendentList;
+  }
+
+  /**
+   * bfs traversal to check path from child->parent
+   */
+  static bool _checkPath(
+      int childNode, int parentNode, Map<int, Node> nodeMap) {
+    Queue<int> nodeQueue = new Queue();
+    nodeQueue.add(childNode);
+    nodeMap[childNode]!.visited = true;
+    while (nodeQueue.isNotEmpty) {
+      int nodeAncestor = nodeQueue.removeLast();
+      for (int link in nodeMap[nodeAncestor]!.childrenList) {
+        if (link == parentNode) return true;
+        if (!nodeMap[link]!.visited) {
+          nodeMap[link]!.visited = true;
+          nodeQueue.add(link);
+        }
+      }
+    }
+    nodeMap.forEach((node, value) {
+      value.visited = false;
+    });
+    return false;
+  }
```

Comment:
Good way to do
An easier way would be to just fetch ancestors of the child node and check if the parent exists in that set of ancestors as you've already written functions for getting ancestors (or vice versa with fetching set of children of the parent and searching for the child in the set of children)